plugins/woocommerce/includes/data-stores/class-wc-customer-data-store-session.php

```diff
@@ -81,14 +84,32 @@ public function save_to_session( $customer ) {
 			if ( 'billing_' === substr( $session_key, 0, 8 ) ) {
 				$session_key = str_replace( 'billing_', '', $session_key );
 			}
-			$data[ $session_key ] = (string) $customer->{"get_$function_key"}( 'edit' );
+			if ( 'meta_data' === $session_key ) {
+				// This is an allow-list of meta data keys which we want to store in session. Avoids storing everything for logged in users.
+				$allowed_meta_data_keys = array(
+					CheckoutFields::ADDITIONAL_FIELDS_KEY,
+					CheckoutFields::BILLING_FIELDS_KEY,
+					CheckoutFields::SHIPPING_FIELDS_KEY,
+				);
+				$session_value          = wp_json_encode(
+					array_filter(
+						$customer->get_meta_data(),
+						function( $meta_data ) use ( $allowed_meta_data_keys ) {
+							return in_array( $meta_data->key, $allowed_meta_data_keys, true );
+						}
+					)
+				);
+			} else {
+				$session_value = $customer->{"get_$function_key"}( 'edit' );
+			}
+			$data[ $session_key ] = (string) $session_value;
```

Comment:
I'm afraid this logic is getting a bit too far and touching places it shouldn't? Can we filter out the allowed meta data keys list and move that logic away? we can keep the wp_json_encode here.

Reply:
We could, but what is the benefit to doing so?

Reply:
I'm not sure of the tangible _benefit_, but my gut feeling is that WC Blocks code is creeping in here and doesn't belong. I do think we should have a filter here and then use that in our own code.

Reply:
Needed to make it init during store api requests so it can hook in time, but change here https://github.com/woocommerce/woocommerce/pull/44181/commits/497f7c5bd01fc23896c41f5015e5975b838dbf5a